ginivision/src/main/java/net/gini/android/vision/analysis/AnalysisScreenPresenter.java

```diff
@@ -0,0 +1,482 @@
+    private boolean mAnalysisCompleted;
+
+    AnalysisScreenPresenter(
+            @NonNull final Application app,
+            @NonNull final AnalysisScreenContract.View view,
+            @NonNull final Document document,
+            @Nullable final String documentAnalysisErrorMessage) {
+        this(app, view, document, documentAnalysisErrorMessage,
+                new AnalysisInteractor(app));
+    }
+
+    @VisibleForTesting
+    AnalysisScreenPresenter(
+            @NonNull final Application app,
+            @NonNull final AnalysisScreenContract.View view,
+            @NonNull final Document document,
+            @Nullable final String documentAnalysisErrorMessage,
+            @NonNull final AnalysisInteractor analysisInteractor) {
+        super(app, view);
+        view.setPresenter(this);
+        mMultiPageDocument = asMultiPageDocument(document);
+        // Tag the documents to be able to clean up the automatically parcelled data
+        tagDocumentsForParcelableMemoryCache(document, mMultiPageDocument);
+        mDocumentAnalysisErrorMessage = documentAnalysisErrorMessage;
+        mAnalysisInteractor = analysisInteractor;
+        mHints = generateRandomHintsList();
+    }
+
+    private List<AnalysisHint> generateRandomHintsList() {
+        final List<AnalysisHint> list = AnalysisHint.getArray();
+        Collections.shuffle(list, new Random());
+        return list;
+    }
+
+    private void tagDocumentsForParcelableMemoryCache(
+            @NonNull final Document document,
+            @NonNull final GiniVisionMultiPageDocument<GiniVisionDocument, GiniVisionDocumentError>
+                    multiPageDocument) {
+        if (document instanceof GiniVisionDocument) {
+            ((GiniVisionDocument) document).setParcelableMemoryCacheTag(
+                    PARCELABLE_MEMORY_CACHE_TAG);
+        }
+        multiPageDocument.setParcelableMemoryCacheTag(PARCELABLE_MEMORY_CACHE_TAG);
+    }
+
+    @SuppressWarnings("unchecked")
+    private GiniVisionMultiPageDocument<GiniVisionDocument,
+            GiniVisionDocumentError> asMultiPageDocument(@NonNull final Document document) {
+        if (!(document instanceof GiniVisionMultiPageDocument)) {
+            return DocumentFactory.newMultiPageDocument((GiniVisionDocument) document);
+        } else {
+            return (GiniVisionMultiPageDocument) document;
+        }
+    }
+
+    @Override
+    void finish() {
+        clearParcelableMemoryCache();
+    }
+
+    @VisibleForTesting
+    void clearParcelableMemoryCache() {
+        // Remove data from the memory cache. The data had been added when the document in the
+        // arguments was automatically parcelled when the activity was stopped
+        ParcelableMemoryCache.getInstance().removeEntriesWithTag(PARCELABLE_MEMORY_CACHE_TAG);
+    }
+
+    @Override
+    public void hideError() {
+        getView().hideErrorSnackbar();
+    }
+
+    @Override
+    public void onNoExtractionsFound() {
+
```

Comment:
I will add a comment explaining, that this method is used in the Screen API activity and in the Component API it has no effect, because we don't have to do any UI changes for this.

In the Screen API activity when this is called we either show the no results screen or just finish.